accepted/2020/serializer/WriteableDomAndDynamic.md

```diff
@@ -0,0 +1,587 @@
+- String Enums. A CLR `Enum` is normally serialized to JSON as a number, but through a custom converter (including `JsonStringEnumConverter`), an `Enum` can be serialized as the string representation.
+
+For example, consider this **hypothetical API** that has a `JsonNumber` class (and not the proposal of using just `JsonNode`)
+```cs
+// Ask for a number with a JSON number token type.
+// Internally probably stored as a byte[] or JsonElement.
+JsonNumber number = JsonSerializer.Deserialize<JsonNumber>("1.2");
+double dlb = number.GetValue<double>();
+
+// Change to "NaN".
+number.Value = double.Nan;
+string json = number.Serialize();
+Debug.Assert(json == "\"NaN\""); // Due to quoted number support, this works.
+
+// However what about round-tripping?
+// Can a JSON string be deserialized into a JsonNumber? If so, "JsonNumber" no longer means a "JSON number".
+// When deserializing a string there is no known CLR type, so the internal value would need be based on a string\byte[].
+// So a "JsonString" node would normally be created, not "JsonNumber" and thus throw here:
+number = JsonSerializer.Deserialize<JsonNumber>(json); // Throws
+
+// Even if the above could be made to work, it wouldn't work with a JsonObject node.
+// A JsonObject would create "JsonString" nodes for "NaN" JSON strings since it has
+// no way of knowing these should map to a JsonNumber (or a CLR number).
+```
+
+Thus the DOM is based on CLR types (not JSON tokens) in edit mode and during serialization. The JSON "value" types (number, string, boolean) are represented as a common `JsonValue` sealed class, and **not** separate `JsonNumber`, `JsonString` and `JsonBoolean` classes as shown in the example above.
+
+During serialization, the DOM is written using the serializer and uses any custom converters that may be configured. In all but idiocentric scenarios (e.g. a custom converter can do anything), the behavior is natural and JSON is round-tripped as expected.
+
+In the example, a JSON "NaN" string is deserialized into a `JsonValue` and when `GetValue<double>` is called, a `double` is returned, assuming the appropriate `JsonSerializerOptions` are configured for quoted numbers. When the node is serialized, it produces a JSON string (again, if configured). The consumer of the DOM is not necessarily aware that `double` values assigned to the node may be serialized to JSON as either a string or a JSON number, and after deserialization is not aware what the internal state is - it could be a `byte[]`, a `JsonElement` or a `string` depending on implementation and\or the JSON token (string or number).
+
+## Primary API
+All types live in STJ.dll.
+```cs
+namespace System.Text.Json
+// Also consider existing S.T.J.Serialization namespace instead of S.T.J.
+// since it is now coupled to the serializer rather than stand-alone.
```

Comment:
To me, this doesn't follow the principle of least surprise. I would not have expected the JsonNode/DOM types to have strong coupling with the serializer at all. We should consider keeping the writable dom at the STJ namespace next to JsonDocument and JsonElement.

Reply:
Due to recent additions to interop with `JsonElement` the design now assumes a "System.Text.Json.Node" namespace.